app/services/moderator/merge_user.rb

```diff
@@ -55,45 +58,77 @@ def update_social
       end
     end
 
+    def merge_sponsorships
+      @delete_user.sponsorships.update_all(user_id: @keep_user.id)
+    end
+
     def merge_profile
-      if @delete_user.github_repos.any?
-        @delete_user.github_repos.update_all(user_id: @keep_user.id)
-        @keep_user.touch(:github_repos_updated_at)
-      end
-      if @delete_user.badge_achievements.any?
-        @delete_user.badge_achievements.update_all(user_id: @keep_user.id)
-        BadgeAchievement.counter_culture_fix_counts(where: { users: { id: @keep_user.id } })
-      end
+      @delete_user.github_repos.update_all(user_id: @keep_user.id)
+      @keep_user.touch(:github_repos_updated_at)
 
       @keep_user.update_columns(created_at: @delete_user.created_at) if @delete_user.created_at < @keep_user.created_at
     end
 
-    def merge_chat_mentions
-      any_memberships = @delete_user.chat_channel_memberships.any?
-      @delete_user.chat_channel_memberships.update_all(user_id: @keep_user.id) if any_memberships
-      @delete_user.mentions.update_all(user_id: @keep_user.id) if @delete_user.mentions.any?
+    def merge_badge_achievements
+      duplicate_badges = @delete_user.badge_achievements.pluck(:badge_id).intersection(
+        @keep_user.badge_achievements.pluck(:badge_id),
+      )
+      @delete_user.badge_achievements.where(badge_id: duplicate_badges).delete_all
+      @delete_user.badge_achievements.update_all(user_id: @keep_user.id)
```

Comment:
Okay, here is the code that I don't feel that great about... This code pattern is what I use to handle duplicates. The steps are:
1. Find the (scoped) intersecting rows
2. Delete the intersecting rows that belong to the user-to-be-deleted
3. Update the rest of the non-conflicting rows from user-to-be-deleted and transfer it to user-to-be-kept

This pattern is 4 queries total (pseudo SQL), and more for scoped associations:
- `SELECT association_id FROM table WHERE user_id = delete-user-id;` 1
- `SELECT association_id FROM table WHERE user_id = keep-user-id;` 2
- `duplicate_associations = intersection_of_two_queries` (just in memory, not a query)
- `DELETE FROM table WHERE association_id = duplication_associations AND user_id = delete-user-id;` 3
- `UPDATE FROM table WHERE user_id = delete-user-id SET user_id = keep-user-id;`

For `BadgeAchievements`, there isn't a scope and is 4 queries, but for `Follows`, `Reactions`, and `ChatChannelMemberships` those are more complicated cases.

I think we could possibly bring it down to 3 queries, but I couldn't figure out how to return only the intersecting values in one query... Maybe others better to querying would like a challenge. 😛

Reply:
similar to my comment about chat channel memberships (where it is more complicated because of the mod role scopes) 

```
@delete_user
  .badge_achievements
  .where(
     badge_id: @keep_user.badge_achievements.select(:badge_id)
  )
  BadgeAchievement Load (1.9ms)  SELECT "badge_achievements".* FROM "badge_achievements" WHERE "badge_achievements"."user_id" = $1 AND "badge_achievements"."badge_id" IN (SELECT "badge_achievements"."badge_id" FROM "badge_achievements" WHERE "badge_achievements"."user_id" = $2)  [["user_id", 4], ["user_id", 2]]
```